python/zfp.pyx

```diff
@@ -0,0 +1,210 @@
+import cython
+from libc.stdlib cimport malloc, free
+import operator
+import functools
+from cython cimport view
+from cpython cimport array
+import array
+
+cimport zfp
+
+import numpy as np
+cimport numpy as np
+
+# export #define's
+HEADER_MAGIC = ZFP_HEADER_MAGIC
+HEADER_META = ZFP_HEADER_META
+HEADER_MODE = ZFP_HEADER_MODE
+HEADER_FULL = ZFP_HEADER_FULL
+
+# export enums
+type_none = zfp_type_none
+type_int32 = zfp_type_int32
+type_int64 = zfp_type_int64
+type_float = zfp_type_float
+type_double = zfp_type_double
+mode_null = zfp_mode_null
+mode_expert = zfp_mode_expert
+mode_fixed_rate = zfp_mode_fixed_rate
+mode_fixed_precision = zfp_mode_fixed_precision
+mode_fixed_accuracy = zfp_mode_fixed_accuracy
+
+
+cpdef dtype_to_ztype(dtype):
+    if dtype == np.int32:
+        return zfp_type_int32
+    elif dtype == np.int64:
+        return zfp_type_int64
+    elif dtype == np.float32:
+        return zfp_type_float
+    elif dtype == np.float64:
+        return zfp_type_double
+    else:
+        raise TypeError("Unknown dtype: {}".format(dtype))
+
+cpdef dtype_to_format(dtype):
+    # format characters detailed here:
+    # https://docs.python.org/2/library/array.html#module-array
+    if dtype == np.int32:
+        return 'i' # signed int
+    elif dtype == np.int64:
+        return 'l' # signed long
+    elif dtype == np.float32:
+        return 'f' # float
+    elif dtype == np.float64:
+        return 'd' # double
+    else:
+        raise TypeError("Unknown dtype: {}".format(dtype))
+
+zfp_to_dtype_map = {
+    zfp_type_int32: np.int32,
+    zfp_type_int64: np.int64,
+    zfp_type_float: np.float32,
+    zfp_type_double: np.float64,
+}
+cpdef ztype_to_dtype(zfp_type ztype):
+    try:
+        return zfp_to_dtype_map[ztype]
+    except KeyError:
+        # TODO: is this the correct type?
+        print "Using np.bytes_"
+        return np.bytes_
+
+cdef size_t sizeof_ztype(zfp_type ztype):
+    if ztype == zfp_type_int32:
+        return sizeof(signed int)
+    elif ztype == zfp_type_int64:
+        return sizeof(signed long long)
```

Comment:
Not sure if Cython assumes any particular C standard, but long long is not C89.  We use some preprocessor magic to provide int32 and int64 types in zfp.  Can those be used instead?

Reply:
:+1: It turns out that the whole function can be replaced with `numpy.dtype`'s `itemsize` method.